src/DynamoCore/Search/SearchDictionary.cs

```diff
@@ -101,8 +101,13 @@ public void Add(V value, IEnumerable<string> tags, double weight = 1)
                 entryDictionary[value] = keys;
                 OnEntryAdded(value);
             }
-            foreach (var tag in tags.Select(x => x.ToLower()))
-                keys[tag] = weight;
+
+            int tagsCount = tags.Count();
+            for (int i = 0; i < tagsCount; i++)
+            {
+                var tag = tags.ElementAt(i).ToLower();
+                keys[tag] = weights.ElementAt(i);
```

Comment:
We should check that these two lists have the same length. If they don't, we should throw a descriptive exception.